### ISSUE TITLE
Assigning user defined properties

### ISSUE THREAD

**Issue by mattheffernan** [MrWolfZ/ngrx-forms#85]
1. Is there a way to assign user defined properties on the initial value (createFormGroupState or something similar?)
2. On a nested form structure, what is the most efficient way to assign different values across the nested form structure but they all have the same property. For example your user def property is 'placeHolder' and each of your fields has an unique value.  I  can do it currently with an updateGroup() per object/form  but it doesn't seem to be very efficient. 

Basically I have a complex form that needs to be extended, and I'm using user def properties to do that.  I'm not user the best way to get that extended data into the form object.  I can do it with a ton of updateGroup calls but it just seems messy.

I did figure out if i recreate the entire initial_value object by hand that would be generated by the updateGroup/createFormGroupState solution I can do it but that seems like a very brittle solution.

**Comment by MrWolfZ** [comment]
The best way to achieve this would indeed be to simply perform an `updateGroup` after creating the initial form state via `createFormGroupState`. To get rid of repeating the 'placeholder' string you can of course create your own custom update function like this:

```typescript
const setPlaceholderProperty = (v: any) => setUserDefinedProperty('placeholder', v);
```

However, if you really do not want to use the `updateGroup` way you can use `updateRecursive` for a somewhat generic approach. `updateRecursive` allows you to apply an update function to every state in the form state tree. By using the fact that **ngrx-forms** builds the ID of a state according to the "path" it takes in the form value you can then assign placeholder values from an object (using an implementation detail like this is something that I don't usually recommend, but this one should really never change). The solution would look something like this (untested, so it may contain bugs):

```typescript
interface FormValue {
  someString: string;
  nested: {
    nestedString: string;
    nestedNumber: number;
  };
}

const INITIAL_FORM_VALUE: FormValue = {
  someString: '',
  nested: {
    nestedString: '',
    nestedNumber: 0,
  },
};

// you could use a mapped type here to get some type safety;
// also, the placeholders object can be partial if not all fields should have one
const PLACEHOLDERS = {
  someString: 'some string',
  nested: {
    nestedNumber: 'nested number',
  },
};

const FORM_ID = 'myForm';

const INITIAL_FORM_STATE = updateRecursive(
  createFormGroupState(FORM_ID, INITIAL_FORM_VALUE),
  s => {
    // this returns the "path" of the state in the form value shape, i.e. the ID
    // of "nestedNumber" is "myForm.nested.nestedNumber"; we take everything but
    // the first part of the path and use the rest to navigate through the
    // placeholders object to get the correct placeholder for the state
    const pathParts = s.id.split('.').slice(1);
    const placeholder = pathParts.reduce((o, p) => o && o[p], PLACEHOLDERS as any);
    return !placeholder ? s : setUserDefinedProperty('placeholder', placeholder, s);
  },
);
```

**Comment by mattheffernan** [comment]
Thanks for the help!  I like the idea of doing the updateRecursive to handle these type of larger merges.
